src/Microsoft.ML.Data/DataLoadSave/Text/TextLoader.cs

```diff
@@ -92,10 +98,18 @@ public Column(string name, DataKind? type, Range[] source, KeyCount keyCount = n
             public string Name;
 
             /// <summary>
-            /// <see cref="DataKind"/> of the items in the column. If <see langword="null"/> defaults to a float.
+            /// <see cref="InternalDataKind"/> of the items in the column. If <see langword="null"/> defaults to a float.
+            /// Although <see cref="InternalDataKind"/> is internal, <see cref="Type"/>'s information can be publically accessed by <see cref="ItemType"/>.
             /// </summary>
             [Argument(ArgumentType.AtMostOnce, HelpText = "Type of the items in the column")]
-            public DataKind? Type;
+            [BestFriend]
+            internal InternalDataKind? Type;
+
+            /// <summary>
+            /// <see cref="DataKind"/> of the items in the column.
+            /// </summary>
+            /// It's a public interface to access the information in an internal DataKind.
+            public DataKind ItemType => Type.HasValue ? Type.Value.ToScalarType() : DataKind.Single;
```

Comment:
Why `ItemType`? Typically we use that name for the type of items inside of something - like a Vector. Why not just `Type` as the public name?

Reply:
We can't. Internal name, nameof(Type), is used in naming command line arguments.

---
In reply to: [259520596](https://github.com/dotnet/machinelearning/pull/2661#discussion_r259520596) [](ancestors = 259520596)

Reply:
I'm not sure I understand this reasoning. Can you explain further?

In general, we shouldn't be sacrificing our public API for "command line" reasonings.

Reply:
We use nameof(Type) as an argument name in entry point (classical APIs) and command line. If we rename `Type` to, for example, `InternalType`, old applications may be broken.

---
In reply to: [259523340](https://github.com/dotnet/machinelearning/pull/2661#discussion_r259523340) [](ancestors = 259523340)

Reply:
I'm still not convinced.... We are breaking all sorts of old applications with the API changes we've been making. We should define our public API as we want, and the other things will need to change.

@TomFinley - thoughts here?

Reply:
TextLoader is a widely used class. Braking BC may cost a lot. Also, `Type` is not very accurate. When we load a float vector column, its type will still be `Float` instead of `Float[]`.

---
In reply to: [259541585](https://github.com/dotnet/machinelearning/pull/2661#discussion_r259541585) [](ancestors = 259541585)

Reply:
Also, we have this design guideline:

https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/constructor

> ✓ DO use the same name for constructor parameters and a property if the constructor parameters are used to simply set the property.

> The only difference between such parameters and the properties should be casing.

So this name should be in sync with the constructor parameter name. Maybe that would solve all your problems?

`public DataKind DataKind` to match the constructor parameter name: `public Column(string name, DataKind dataKind, ...`

Reply:
> TextLoader is a widely used class.

Agreed. That is why it is critical we get the API right. We can never change it after this.

Reply:
Ok. Let's use `DataKind`.

---
In reply to: [259542584](https://github.com/dotnet/machinelearning/pull/2661#discussion_r259542584) [](ancestors = 259542584)